dulwich/graph.py

```diff
@@ -20,11 +20,73 @@
 
 """Implementation of merge-base following the approach of git."""
 
-from collections import deque
-from typing import Deque
+from collections.abc import MutableMapping
+from collections import OrderedDict
 
+from heapq import heappush, heappop
 
-def _find_lcas(lookup_parents, c1, c2s):
+# Because opening an sha file from the object_store, reading it,
+# then decoding its will be a slow operation, we want to create
+# a cache of Commit objects so that repeat parent and commit time
+# lookup (which are inherent when looking for a common ancestor)
+# will be much faster, so we need to implement a simple Commit cache
+# using an OrderedDict
+
+
+class CommitCache(MutableMapping):
+
+    def __init__(self, maxsize):
+        self.maxsize = maxsize
+        self.d = OrderedDict()
+
+    def __getitem__(self, key):
+        self.d.move_to_end(key)
+        return self.d[key]
+
+    def __setitem__(self, key, value):
+        if key in self.d:
+            self.d.move_to_end(key)
+        elif len(self.d) == self.maxsize:
+            self.d.popitem(last=False)
+        self.d[key] = value
+
+    def __delitem__(self, key):
+        del self.d[key]
+
+    def __iter__(self):
+        return self.d__iter__()
+
+    def __len__(self):
+        return len(self.d)
+
+    def __contains__(self, key):
+        return key in self.d
+
+
+# priority queue using builtin python minheap tools
+# why they do not have a builtin maxheap is simply ridiculous but
+# liveable with integer time stamps using negation
+class WorkList(object):
```

Comment:
Please add tests for WorkList specifically.

Reply:
Will do.   FYI worklist has existed in some form or other since the beginning and this version uses built-in python minheap.  The test would be to add 3 nodes out of time order and prove they are ordered properly upon retrieval.

Reply:
Test for WorkList added to test_graph.py in the updated PR